Lib/pathlib.py

```diff
@@ -1310,6 +1310,120 @@ def expanduser(self):
+        subdirectories (directories are generated bottom up).
+
+        When top_down is True, the caller can modify the dirnames list in-place
+        (e.g., via del or slice assignment), and walk will only recurse
+        into the subdirectories whose names remain in dirnames; this
+        can be used to prune the search, or to impose a specific order of
+        visiting. Modifying dirnames when
+        top_down is False has no effect on the behavior of Path.walk(), since the
+        directories in dirnames have already been generated by the time dirnames
+        itself is generated. No matter the value of top_down, the list of
+        subdirectories is retrieved before the tuples for the directory and its
+        subdirectories are generated.
+
+        By default errors from Path._scandir() call are ignored.  If
+        optional arg 'on_error' is specified, it should be a callable; it
+        will be called with one argument, an OSError instance.  It can
+        report the error to continue with the walk, or raise the exception
+        to abort the walk.  Note that the filename is available as the
+        filename attribute of the exception object.
+
+        By default, Path.walk does not follow symbolic links to subdirectories
+        on systems that support them.  In order to get this functionality, set
+        the optional argument 'follow_symlinks' to true. Unlike os.walk,
+        Path.walk only adds symbolic links to dirnames if follow_symlinks=True. 
+
+        Caution: if self is a relative Path, don't change the
+        current working directory between resumptions of walk. walk never
+        changes the current directory, and assumes that the caller doesn't
+        either.
+        
+        Caution: Unlike os.walk, Path.walk assumes the directories have not
+        been modified between its resumptions. I.e. If a directory from
+        dirnames has been replaced with a symlink and follow_symlinks=False,
+        walk will still try to descend into it. To prevent such behavior,
+        remove directories from dirnames if they have been modified and you
+        do not want Path.walk to descend into them anymore.
+
+        Example:
+
+        from pathlib import Path
+        for root, dirs, files in Path().walk(on_error=print):
+            print(
+                root,
+                "consumes",
+                sum((root / file).stat().st_size for file in files),
+                "bytes in",
+                len(files),
+                "non-directory files"
+            )
+            # don't visit __pycache__ directories
+            if '__pycache__' in dirs:
+                dirs.remove('__pycache__')
+        """
+        sys.audit("pathlib.Path.walk", self, on_error, follow_symlinks)
+        return self._walk(top_down, on_error, follow_symlinks)
+
+    def _walk(self, top_down, on_error, follow_symlinks):
+        dirnames = []
+        filenames = []
```

Comment:
Move these to just before `with scandir_it:`. This avoids an allocation if `self._scandir()` raises an exception.